webapp/channels/src/workers/sw.test.js

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+describe('Service Worker', () => {
+    let sw;
+
+    beforeEach(() => {
+        // Mock the ServiceWorker global scope
+        global.caches = {
+            open: jest.fn(() => Promise.resolve({
+                put: jest.fn(),
+            })),
+            match: jest.fn(),
+            keys: jest.fn(),
+            delete: jest.fn(),
+        };
+        global.fetch = jest.fn();
+
+        // Import the service worker fresh for each test
+        jest.isolateModules(() => {
+            sw = require('./sw.js');
+        });
+    });
+
+    afterEach(() => {
+        jest.resetModules();
+    });
+
+    describe('putInCache', () => {
+        it('should store HTML documents under HTML_REQUEST key', async () => {
+            const cache = {put: jest.fn()};
+            global.self.caches.open.mockResolvedValue(cache);
+
+            const request = {destination: 'document'};
+            const response = {clone: () => response};
+
+            await sw.putInCache(request, response);
+
+            expect(global.self.caches.open).toHaveBeenCalledWith(sw.cacheName);
+            expect(cache.put).toHaveBeenCalledWith(sw.HTML_REQUEST, response);
+        });
+
+        it('should store non-HTML requests under their own key', async () => {
+            const cache = {put: jest.fn()};
+            global.self.caches.open.mockResolvedValue(cache);
+
+            const request = {destination: 'image'};
+            const response = {clone: () => response};
+
+            await sw.putInCache(request, response);
+
+            expect(global.self.caches.open).toHaveBeenCalledWith(sw.cacheName);
+            expect(cache.put).toHaveBeenCalledWith(request, response);
+        });
+    });
+
+    describe('handleCacheRequest', () => {
+        it('should return cached response if available', async () => {
+            const cachedResponse = {clone: () => cachedResponse};
+            global.self.caches.match.mockResolvedValue(cachedResponse);
+
+            const networkResponse = {clone: () => networkResponse};
+            global.fetch.mockResolvedValue(networkResponse);
+
+            const request = {destination: 'style', clone: () => request};
+            const result = await sw.handleCacheRequest(request);
+
+            expect(result).toBe(cachedResponse);
+        });
```

Comment:
I don't know if I'm missing something because this is new to me, but according to the code, this should never happen, right? The request's destination is `style`, so it should be retrieved from cache only if its `url` matches one of the `precachedAssets`, but this request's `url` is `undefined`.

Reply:
Haha this is happening because cachedResponse and networkResponse are the same objects. I'll update it.

Reply:
Oh wait, this is correct. Because we are mocking the `caches.match` function to return a cached response.

Reply:
It's correct because we're mocking, but it doesn't follow what the code does, right? In the real code, a request with destination `style` with an `undefined` URL should never return a cached response. I'd be happy if we change the test so that the destination is `document`, I guess :)

Reply:
Yeah we are mocking it. Tbh, I can remove the test entirely because it's not really doing much. There's too much mocking going on. LMK.

Reply:
I mean, I agree it's not much, but it's testing *something*. My main point is that the mocking is wrong: if `destination` equals `style`, and the URL of the request is not in the list of cached URLs, then `global.self.caches.match` should return `undefined`. And so the result should be the network response.

This is probably not worth all this time, though :joy: So I'll let you decide next steps here :)